src/streams/register_nack.rs

```diff
@@ -0,0 +1,560 @@
+                last.blp |= 1 << pos;
+            } else {
+                nacks.push(NackEntry {
+                    pid: (seq % U16_MAX) as u16,
+                    blp: 0,
+                });
+                last_seq_added = seq;
+            }
+
+            self.packet(seq.into()).nack_count += 1;
+        }
+
+        if nacks.is_empty() {
+            return None;
+        }
+
+        let reports = ReportList::lists_from_iter(nacks).into_iter();
+
+        Some(
+            reports
+                .map(|reports| {
+                    Nack {
+                        sender_ssrc: 0.into(),
+                        ssrc: 0.into(), // changed when sending
+                        reports,
+                    }
+                })
+                .collect(),
+        )
+    }
+
+    fn as_index(&self, seq: SeqNo) -> usize {
+        (*seq % self.packets.len() as u64) as usize
+    }
+
+    fn packet(&mut self, seq: SeqNo) -> &mut PacketStatus {
+        let index = self.as_index(seq);
+        &mut self.packets[index]
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::ops::Range;
+
+    use crate::streams::register_nack::MAX_MISORDER;
+
+    use super::NackRegister;
+
+    fn assert_update(
+        reg: &mut NackRegister,
+        seq: u64,
+        expect_new: bool,
+        expect_received: bool,
+        expect_active: Range<u64>,
+    ) {
+        assert_eq!(
+            reg.update(seq.into()),
+            expect_new,
+            "seq {} was expected to{} be new",
+            seq,
+            if expect_new { "" } else { " NOT" }
+        );
+        let active = reg.active.clone().expect("nack range");
+        assert_eq!(
+            reg.packet(seq.into()).received,
+            expect_received,
+            "seq {} expected to{} be received in {:?}",
+            seq,
+            if expect_received { "" } else { " NOT" },
+            active
+        );
+        assert_eq!(active, expect_active.start.into()..expect_active.end.into());
+        assert_not_dirty(reg);
+    }
+
+    fn assert_not_dirty(reg: &NackRegister) {
+        // we should leave no dirty state outside of the nack window
+        let active = reg.active.clone().expect("nack range");
+        let active = (*active.start..=*active.end)
+            .map(|seq| reg.as_index(seq.into()))
+            .collect::<Vec<_>>();
+
+        for i in 0..reg.packets.len() {
+            if active.contains(&i) {
+                continue;
+            }
+            assert!(
+                !reg.packets[i].received && reg.packets[i].nack_count == 0,
+                "dirty state at index {} outside of nack window {:?}",
+                i,
+                active,
+            );
+        }
+    }
+
+    #[test]
+    fn active_window_sliding() {
+        let mut reg = NackRegister::new();
+
+        assert_update(&mut reg, 10, true, true, 10..10);
+
+        // packet before window start is ignored
+        assert_update(&mut reg, 9, false, false, 10..10);
```

Comment:
Does this means we could lose out on some packets if we get unlucky with re-ordering at the start of an RTP stream? i.e. if packets 11, 10, 9 arrive in quick succession in that order we'd ignore 10 and 9. At the start of a video stream these are also likely to be part of a keyframe so their loss is particularly problematic.

Reply:
Correct

Reply:
This was already the case, so it's not better or worse.

Reply:
Is it something we want to address long term?

Reply:
Could we detect and `warn!` to see how much better it'd be?

Reply:
@k0nserv we would not ignore packets as you wrote. The comment there is from the perspective of nacking. We would not consider them for nacking. 
So `11, 10, 9` is fine (no loss). An issue would be if we receive `11, 9` at the start because we would not nack `10`. However even if we received `11, 10`, we would not nack `9` because .. how far back should we go ?

Reply:
yes. also what i said before is not 100% accurate, because we also don't flush them out in RTP mode (based on the fact that the register does not register them as new). 

Needs more thought.

Reply:
I have made it so that the window extends back if the packet is within` MAX_MISORDER`, or `warn!` otherwise

Reply:
There are issues with this, I have ideas how to solve but I put it in a separate PR.